azure/services/virtualmachineimages/images.go

```diff
@@ -0,0 +1,167 @@
+}
+
+// GetDefaultWindowsImage returns the default image spec for Windows.
+func GetDefaultWindowsImage(ctx context.Context, auth azure.Authorizer, location, k8sVersion, runtime, osAndVersion string) (*infrav1.Image, error) {
+	v122 := semver.MustParse("1.22.0")
+	v, err := semver.ParseTolerant(k8sVersion)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse Kubernetes version \"%s\"", k8sVersion)
+	}
+
+	// If containerd is specified we don't currently support less than 1.22
+	if v.LE(v122) && runtime == "containerd" {
+		return nil, errors.New("containerd image only supported in 1.22+")
+	}
+
+	if osAndVersion == "" {
+		osAndVersion = azure.DefaultWindowsOsAndVersion
+	}
+
+	publisher, offer := azure.DefaultImagePublisherID, azure.DefaultWindowsImageOfferID
+	skuID, version, err := getDefaultImageSKUIDAndVersion(
+		ctx, auth, location, publisher, offer, k8sVersion, osAndVersion)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get default image")
+	}
+
+	// Starting with 1.22 we default to containerd for Windows unless the runtime flag is set.
+	if v.GTE(v122) && runtime != "dockershim" {
+		skuID += "-containerd"
+	}
+
+	defaultImage := &infrav1.Image{
+		Marketplace: &infrav1.AzureMarketplaceImage{
+			Publisher: publisher,
+			Offer:     offer,
+			SKU:       skuID,
+			Version:   version,
+		},
+	}
+
+	return defaultImage, nil
+}
+
+// k8sVersionInSKUName returns true if the k8s version is in the SKU name (the older style of naming).
+func k8sVersionInSKUName(major, minor, patch uint64) bool {
+	return (major == 1 && minor < 21) ||
+		(major == 1 && minor == 21 && patch <= 12) ||
+		(major == 1 && minor == 22 && patch <= 9) ||
+		(major == 1 && minor == 23 && patch <= 6)
+}
+
+// GetDefaultImageSKUID gets the SKU ID and version of the image to use for the provided version of Kubernetes.
+// note: osAndVersion is expected to be in the format of {os}-{version} (ex: ubuntu-2004 or windows-2022)
+func getDefaultImageSKUIDAndVersion(ctx context.Context, auth azure.Authorizer, location, publisher, offer, k8sVersion, osAndVersion string) (string, string, error) {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "virtualmachineimages.AzureClient.getDefaultImageSKUIDAndVersion")
+	defer done()
+
+	v, err := semver.ParseTolerant(k8sVersion)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "unable to parse Kubernetes version \"%s\" in spec, expected valid SemVer string", k8sVersion)
+	}
+
+	// Old SKUs before 1.21.12, 1.22.9, or 1.23.6 are named like "k8s-1dot21dot2-ubuntu-2004".
+	if k8sVersionInSKUName(v.Major, v.Minor, v.Patch) {
+		return fmt.Sprintf("k8s-%ddot%ddot%d-%s", v.Major, v.Minor, v.Patch, osAndVersion), azure.LatestVersion, nil
+	}
+
+	// New SKUs don't contain the Kubernetes version and are named like "ubuntu-2004-gen1".
+	sku := fmt.Sprintf("%s-gen1", osAndVersion)
+	// Look up the version ID for the given SKU.
+	listVMImagesResource, err := NewClient(auth).List(ctx, location, publisher, offer, sku)
```

Comment:
thoughts about structuring the image list as a cache (we might even be able to reuse the existing SKUs cache to make it more generic and include images), similar to what we do for [VM SKUs](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/resourceskus/cache.go#L107) so that we can at least ensure we make this expensive call max once per reconcile? And then later we can see how we can cache stuff across multiple reconcile loops, but at least it will be easier since the cache will already be in place.

Reply:
Yes, it needs to be cached for sure. I'm finishing the tricky unit test refactoring and then adding a cache will be the next step.